common/kvstore/tablestore/table_store_builder.go

```diff
@@ -0,0 +1,402 @@
+// modifySchema is false, the tables are loaded as is, and any tables in the provided list are ignored.
+func start(
+	logger logging.Logger,
+	base kvstore.Store,
+	modifySchema bool,
+	tables ...string) (kvstore.TableStore, error) {
+
+	metadataTable := newTableView(base, "metadata", metadataTableID)
+	namespaceTable := newTableView(base, "namespace", namespaceTableID)
+
+	err := validateSchema(metadataTable)
+	if err != nil {
+		return nil, fmt.Errorf("error validating schema: %w", err)
+	}
+
+	tableIDMap, err := loadNamespaceTable(namespaceTable)
+	if err != nil {
+		return nil, fmt.Errorf("error loading namespace table: %w", err)
+	}
+
+	err = handleIncompleteDeletion(logger, base, metadataTable, namespaceTable)
+	if err != nil {
+		return nil, fmt.Errorf("error handling incomplete deletion: %w", err)
+	}
+
+	if modifySchema {
+		err = addAndRemoveTables(base, metadataTable, namespaceTable, tableIDMap, tables)
+		if err != nil {
+			return nil, fmt.Errorf("error adding and removing tables: %w", err)
+		}
+	}
+
+	tableMap := make(map[string]kvstore.Table, len(tableIDMap))
+	for tableID, tableName := range tableIDMap {
+		tableMap[tableName] = newTableView(base, tableName, tableID)
+	}
+	return newTableStore(logger, base, tableMap), nil
+}
+
+// buildBaseStore creates a new base store of the given type.
+func buildBaseStore(storeType StoreType, logger logging.Logger, path string) (kvstore.Store, error) {
+	switch storeType {
+	case LevelDB:
+		return leveldb.NewStore(logger, path)
+	case MapStore:
+		return mapstore.NewStore(), nil
+	default:
+		return nil, fmt.Errorf("unknown store type: %d", storeType)
+	}
+}
+
+// validateSchema loads/initiates the schema version in the metadata table.
+func validateSchema(metadataTable kvstore.Table) error {
+
+	schemaKey := []byte(metadataSchemaVersionKey)
+	onDiskSchemaBytes, err := metadataTable.Get(schemaKey)
+
+	if err != nil {
+		if !errors.Is(err, kvstore.ErrNotFound) {
+			return err
+		}
+
+		// This store is new, no on disk schema version exists.
+		onDiskSchemaBytes = make([]byte, 8)
+		binary.BigEndian.PutUint64(onDiskSchemaBytes, currentSchemaVersion)
+
+		err = metadataTable.Put(schemaKey, onDiskSchemaBytes)
+		if err != nil {
+			return fmt.Errorf("error setting schema version in metadata table: %w", err)
+		}
+
+		return nil
+	}
+
+	// Verify schema version.
+	onDiskSchema := binary.BigEndian.Uint64(onDiskSchemaBytes)
+	if onDiskSchema != currentSchemaVersion {
+		// In the future if we change schema versions, we may need to write migration code here.
+		return fmt.Errorf(
+			"incompatible schema version: code is at version %d, data on disk is at version %d",
+			currentSchemaVersion, onDiskSchema)
+	}
+
+	return nil
+}
+
+// This method adds and removes tables as needed to match the given list of tables. The table ID map is updated
+// to reflect the new state of the tables.
+func addAndRemoveTables(
```

Comment:
Potentially more readability if we build this into:
- computeSchemaChanges: returns tablesToAdd and tablesToDrop
- addTables(tablesToAdd)
- dropTables(tablesToDrop)

Reply:
done